hybrid/compartment_examples/inter_comp_call/switch_compartment.S

```diff
@@ -72,54 +78,14 @@ switch_compartment:
 
     ret       clr
 
-    // Inner helper for cleaning capabilities from registers, either side of an
-    // AAPCS64 function call where some level of distrust exists between caller
-    // and callee.
-    //
-    // Depending on the trust model, this might not be required, but the process
-    // is included here for demonstration purposes. Note that if data needs to
-    // be scrubbed as well as capabilities, then NEON registers also need to be
-    // cleaned.
-    //
-    // Callers should enter at an appropriate offset so that live registers
-    // holding arguments and return values (c0-c7) are preserved.
-clean:
-    mov x0, #0
-    mov x1, #0
-    mov x2, #0
-    mov x3, #0
-    mov x4, #0
-    mov x5, #0
-    mov x6, #0
-    mov x7, #0
-    mov x8, #0
-    mov x9, #0
-    mov x10, #0
-    mov x11, #0
-    mov x12, #0
-    mov x13, #0
-    mov x14, #0
-    mov x15, #0
-    mov x16, #0
-    mov x17, #0
-    // x18 is the "platform register" (for some platforms). If so, it needs to
-    // be preserved, but here we assume that only the lower 64 bits are
-    // required.
-    mov x18, x18
-    // x19-x29 are callee-saved, but only the lower 64 bits.
-    mov x19, x19
-    mov x20, x20
-    mov x21, x21
-    mov x22, x22
-    mov x23, x23
-    mov x24, x24
-    mov x25, x25
-    mov x26, x26
-    mov x27, x27
-    mov x28, x28
-    mov x29, x29  // FP
-    // We need LR (x30) to return. The call to this helper already cleaned it.
-    // Don't replace SP; this needs special handling by the caller anyway.
-    ret
+// Select the appropriate `clean` implementation. Some examples use a correct
+// `clean` implementation, whereas `secure-update_ddc` uses one that
+// intentionally fails to clear `x20`.
+#ifdef INTER_COMP_LEAK_DDC
```

Comment:
Similar to another comment above, there is a dependency here from within a specific example (namely this `INTER_COMP_LEAK_DDC` variable). But perhaps the readability is improved by the two distinct header file names.

Reply:
Good point, though now I'm a bit unsure whether conditionally including a `clean` implementation is such a good idea. As you pointed out, it does hurt readability.

Maybe it would be better to have two `clean` functions (`clean_all` and `clean_leak_x20`): the compartment switcher can just choose one or the other based on the value of a (new) argument passed by the caller (i.e. `switch_compartment` receives an additional `leak_x20` arg).

Reply:
Some thoughts:
* Could lower the `#define` on the specific register we leak. That might allow to lift `clean` back up into `switch_compartment`, if you prefer;
* We could forego entirely the cleaning in the example where we want to leak. Should still keep the essence of what we want to showcase (i.e., capability leak at the register level), without fumbling about with conditionals;
* Two distinct functions are not bad, but there might be a better way than having to pass "state" (i.e., whether we want to leak or not) around. It might be easier for me to evaluate the idea with a concrete example, if you choose this path.

Reply:
> Could lower the #define on the specific mov instruction for the register we want leak, inside clean itself. That might allow to lift clean back up into switch_compartment, if you prefer (although would require preprocessing);

Good point, though I think want to get rid of preprocessing altogether: ideally, all examples should assemble `switch_compartment.S` and link it in the final binary. Currently, the examples `#include` the top-level `switch_compartment.S` from their individual `shared.S`s. In most cases the `#include` isn't necessary (it could just as well be compiled separately and then linked in). However, this conditional include to select the appropriate `clean` makes this difficult to get rid of (someone needs to define `INTER_COMP_LEAK_DDC`: it's either going to be a `#define` preceding the `#include switch_compartment.S`, or a flag passed to the assembler. I think both these options would be less readable than removing preprocessing and adding a `leak_x20` arg).

> We could forego entirely the cleaning in the example where we want to leak. Should still keep the essence of what we want to showcase (i.e., capability leak at the register level), without fumbling about with conditionals;

True, but we'd still need a conditional to skip calling `clean` in `switch_compartment`

> Two distinct functions are not bad, but there might be a better way than having to pass "state" (i.e., whether we want to leak or not) around. It might be easier for me to evaluate the idea with a concrete example, if you choose this path.

Let me try to come up with an example.

Reply:
Those were just some ideas that might help; feel free to commit what you think is best, and we can continue iterating over that, rather than preparing examples and so.

Reply:
I pushed the other implementation option in 2edd256406a0480c26bb8051a33c7ba53fdc755a

Reply:
Since https://github.com/capablevms/cheri-examples/commit/2edd256406a0480c26bb8051a33c7ba53fdc755a is a fairly large commit, can you be more specific where I have to look for these related changes?